### ISSUE TITLE
Investigate whether script block on layout reflow can be removed

### ISSUE THREAD

**Issue by gterzian** [servo/servo#22621]
https://github.com/servo/servo/blob/3ebd42caa3c722e8b0d7cdd7f4a46e9fd7668feb/components/script/dom/window.rs#L1432

```
HangProfile backtrace:
                         - __ZN2js3jit16CallNativeGetterEP9JSContextN2JS6HandleIP10JSFunctionEENS4_IP8JSObjectEENS3_13MutableHandleINS3_5ValueEEE
                         - __ZN6script3dom8bindings5utils12generic_call17he3d4586dd7542919E
                         - __ZN6script3dom8bindings7codegen8Bindings26CSSStyleDeclarationBinding26CSSStyleDeclarationBinding11get_display17h4dd6fdfbf1393160E
                         - ___rust_maybe_catch_panic
                         - __ZN3std9panicking3try7do_call17h14be674bb4719badE.llvm.14731053362047269330
                         - __ZN101_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$9call_once17hd77692f3928353f7E
                         - __ZN6script3dom19cssstyledeclaration19CSSStyleDeclaration18get_property_value17hcefcd3a7974fb185E.llvm.8309293462450537392
                         - __ZN6script3dom6window6Window6reflow17h14e0ce770caf642eE
                         - __ZN6script3dom6window6Window12force_reflow17h79f45d1e1d5f404bE
                         - __ZN54_$LT$crossbeam_channel..channel..Receiver$LT$T$GT$$GT$4recv17hd5fc7a74e53110eaE
                         - __ZN59_$LT$crossbeam_channel..flavors..list..Channel$LT$T$GT$$GT$4recv17h84da2d1a95ba86f2E
                         - _swtch_pri
```

The script-thread seems to quite reliably get into "permanent hangs"(currently defined as a 5 sec block) when waiting for the answer from layout doing a reflow. (above backtrace collected from https://www.nytimes.com/)

Could the block on Layout be replaced with a non-blocking workflow involving layout sending a message back to script, which would be handled as part of the "normal" event-loop?

The logic run using the response from layout seems to be:
https://github.com/servo/servo/blob/3ebd42caa3c722e8b0d7cdd7f4a46e9fd7668feb/components/script/dom/window.rs#L1452

and 

https://github.com/servo/servo/blob/3ebd42caa3c722e8b0d7cdd7f4a46e9fd7668feb/components/script/dom/window.rs#L1486

If this can be moved to logic run in response to the script-thread receiving a message on it's main event-loop, we can remove this block I guess...

**Comment by jdm** [comment]
In the current model, the risk of doing style recalculation in parallel with normal script operation is the risk of mutating the DOM tree and/or GCing nodes that are being laid out. To fix this, some kind of copy-on-write or transactional DOM would be needed, which are both very complex solutions.

**Comment by emilio** [comment]
Yeah, no way style recalc can run in parallel with DOM. Layout could in theory, if we forbid access to nodes and such. I think that's feasible, though images and such are the worst offenders.

**Comment by gterzian** [comment]
In the design wiki I came across this:

> Under most any design with concurrent JS and layout, JS is going to be waiting to query layout sometimes, perhaps often. This will be the most opportune time to run the GC.

Perhaps we could do something like:

1. Send the reflow request.
2. GC
3. Block on the reflow reply

**Comment by emilio** [comment]
That doesn't work if the GC can move nodes.
